src/FSharp.Core/seq.fs

```diff
@@ -875,6 +875,7 @@ module Seq =
         | :? ('T[]) as a -> a.Length = 0
         | :? ('T list) as a -> a.IsEmpty
         | :? ICollection<'T> as a -> a.Count = 0
+        | :? string as a -> a.Length = 0
```

Comment:
That'll likely not work, since it will be constraining type to be string

Reply:
Ah, you're right. My bad, did the edits directly in GitHub because I thought they were so simple.

Is there any way to accomplish this, or does the F# type system simply not allow it?

Reply:
Probably not without boxing it.

You technically can do it in `FSharp.Core` by having different types in `when` cases (special feature restricted to core library). Hoever I am not sure if we want to do it here.

Reply:
> Hoever I am not sure if we want to do it here.

Maybe only if that would show significant improvements in benchmarks. 🙂

Reply:
> Ah, you're right. My bad, did the edits directly in GitHub because I thought they were so simple.
> 
> Is there any way to accomplish this, or does the F# type system simply not allow it?

You could do the check within a 'when' clause, and then downcast. But this should be then re-benchmarked of being worth it, and also not regressing other cases. Something like:

``when source.GetType() = typeof<string> -> (source :?> string).Length = 0``

Best would be to check if/how the condition can be fully erased when 'T is different from char, and avoid boxing.
e.g. prepending the condition with a condition based on 'T:
`` typeof<'T> = typeof<char> &&`` , in a way which JIT understands.

Reply:
No, I didn't mean runtime checks, but the `FSharp.Core`-only `when feature`, like we do with comparison:
```fsharp
let inline GenericLessThanFast (x:'T) (y:'T) = 
	GenericLessThanIntrinsic x y
	when 'T : bool   = (# "clt" x y : bool #)
	when 'T : sbyte  = (# "clt" x y : bool #)
	when 'T : int16  = (# "clt" x y : bool #)
	when 'T : int32  = (# "clt" x y : bool #)
	when 'T : int64  = (# "clt" x y : bool #)
```

Reply:
This is unknown to me. Feel free to do this! My benchmark indicates that there is much to be gained using string-optimized `length` and `isEmpty`, but unfortunately I have too many other obligations at the moment to delve into this.